Refactor to SystemTransitionAnimator

SD10 opened the pull request:
We can continue the discussion of Pull Request #426 here

I was able to condense all the SystemTransitionAnimators into a single class
`SystemTransitionAnimator`

There's probably a better way to do this and I would like to do something similar
for the other TransitionAnimators, but when trying to do everything at once it got crazy fast 😅 

I would also like to condense the `ActivityIndicatorAnimators` if possible.

In short, reduce the object graph and any repeated code as much as possible.

I consider this Pull Request incomplete, I'd just like some feedback before taking it any further.

I also removed the `switch` statements as discussed in #426

SD10 committed:
Add matchingGesture and opposingGesture properties

SD10 committed:
Refactor to SystemTransitionAnimator

tbaranes commented:
I need more time to go this, that's a big one! It looks impressive, good idea to condensate the SystemAnimation.

>  I would like to do something similar
for the other TransitionAnimators, but when trying to do everything at once it got crazy fast

We may improve the animator protocol, but otherwise, we risk to loose readability if we merge them all since the animations are quite different following the cases.

> I would also like to condense the ActivityIndicatorAnimators if possible.

Looking forward for this 😬

tbaranes commented on IBAnimatable/AnimatorFactory.swift:
```diff
@@ -12,11 +12,11 @@ public struct AnimatorFactory {
     switch transitionAnimationType {
     case .systemRotate:
-      return SystemRotateAnimator(transitionDuration: transitionDuration)
+      return SystemTransitionAnimator(systemType: .rotate, duration: transitionDuration)
```
I'm wondering if we can't simplify this switch for the system animation, maybe something like:

```swift
if let systemType = transitionAnimationType.systemType {
   SystemTransitionAnimator(systemType: transitionAnimationType, duration: transitionDuration)
   return // or making an else and splitting in two methods
}

switch transitionAnimationType {
case .portal:
   ...
case .othersCustomAnimator:
   ...
default:
   fatalError("No animator available for this transition.")
}
```

where ` transitionAnimationType.systemType` would be a simple mapping.

A second solution could be to just detect if the `transitionAnimationType` is a system animation or not, then let `SystemTransitionAnimator` makes the mapping job.

SD10 replied:
Great idea, I'll look into this 👍

tbaranes commented on IBAnimatable/FadeAnimator.swift:
```diff
@@ -20,12 +20,9 @@ public class FadeAnimator: NSObject, AnimatedTransitioning {
+    case .in, .out:
+      self.transitionAnimationType = .fade(direction: direction)
+      self.reverseAnimationType = .fade(direction: direction.opposite)
     default:
```
Is this switch still mandatory? 
If it's to be sure that we are using a compatible direction for a fade, I would be in favour to add a `fatalError`:

```swift
if direction != .in && direction != .out && direction != .cross {
   fatalError("Invalid direction for a fade transition")
}
```

SD10 replied:
Since `opposite` for the `.cross` case just returns `.cross` I think it can be removed completely 👍

JakeLin replied:
@SD10 will be great to kill the `switch`

tbaranes commented on IBAnimatable/SystemTransitionAnimator.swift:
```diff
@@ -0,0 +1,97 @@
+    transitionContext.containerView.addSubview(toView)
+
+    performSystemTransition(using: transitionContext)
+
```
Lot of newlines here

SD10 commented:
@tbaranes 
Yeah, I agree this is a hacky solution and not what I intended. 

I too am most concerned with the complexity SystemTransitionAnimator `initializer` 
and maintaining the readability of the code base.

I hope by working on the rest of the `TransitionAnimators` I will come about the final
solution whether it be:
1. having a `TransitionAnimator` base class(es)
2. improving the `AnimatedTransitioning` protocol

I think the dependency on the current enum setup is what has my hands tied, but would need some significant time to investigate decoupling them a bit.

I believe it's a step in the right direction, though.

tbaranes commented on IBAnimatable/TransitionAnimationType.swift:
```diff
@@ -35,6 +35,75 @@ public enum TransitionAnimationType {
+extension TransitionAnimationType {
+
+  /// Reverses the direction of the animation
+  var reversed: TransitionAnimationType {
```
If I didn't miss something, the `reversed` is just used for the `SystemAnimator`, shouldn't we move it in the `SystemTransitionType`?

SD10 replied:
`reversed` reverses the direction of a `TransitionAnimationType` which the `SystemTransitionAnimator` has -- very confusing I know. That's why if I was able to decouple or re-think these enums a bit I think it would be very useful. However, that's no easy task because the String `initializer` is the most important part of the library. Must tread slowly and with caution.

JakeLin replied:
I think most of the `TransitionAnimation` have `reversed` except for `.systemRotate, .systemSuckEffect, .systemRippleEffect, .explode`

tbaranes commented on IBAnimatable/SystemTransitionAnimator.swift:
```diff
@@ -0,0 +1,97 @@
+    self.transitionDuration = duration
+
+    if let direction = direction {
+      self.transitionAnimationType = TransitionAnimationType(systemType: systemType, direction: direction)
```
We are mapping a `TransitionAnimationType` to a `SystemTransitionType ` then rolling back to `TransitionAnimationType`. There's room for improvements here 🤔

SD10 replied:
Totally agree 👍

tbaranes commented:
My apologies, I clicked on "approved" while I just wanted to add a few comments 🙄

SD10 commented:
@tbaranes No problem, I'm nowhere near happy with it myself. I was just looking for some feedback and the "ok" to continue investigating as these are major changes. Your input has already helped me return to the drawing board. There's definitely room for improvement here.

tbaranes commented:
Good job on this, it's already a good first move in order to improve that module 👍 

I'm joining you about the enum, hard to improve without breaking it since it was complex from the beginning. I think we can go with this refactor for now since it improves / simplify the codebase, then going farther by introducing breaking changes if necessary.

Will continue to review it tomorrow 😴

JakeLin commented on IBAnimatable/CardsAnimator.swift:
```diff
@@ -20,14 +20,8 @@ public class CardsAnimator: NSObject, AnimatedTransitioning {
-      self.reverseAnimationType = .cards(direction: .backward)
-    }
+    self.transitionAnimationType = .cards(direction: direction)
+    self.reverseAnimationType = .cards(direction: direction.opposite)
```
Well, much better 🎉

SD10 committed:
Clean up SystemTransitionAnimator initializer

SD10 commented:
@tbaranes I split the initializer for `SystemTransitionAnimator` and it's much cleaner.
I must've been sleep deprived when I was trying to do it in one.
I also removed some of the unnecessary helper methods I created.
Haven't had much time to look further into this yet but hopefully this week.

JakeLin reviewed (changes requested):
@SD10 you are the hero for removing a bunch of code. Well done. 👍

Because it replaced all system transition animator with `SystemTransitionAnimator`, please update the CHANGELOG.md file. Apart from that, it is all good.

SD10 commented:
@JakeLin

In the case of `.systemSuckEffect`, `.systemRippleEffect`, `.systemRotate`, and `.explode`
I just returned the original `TransitionAnimationType`
Would you like me to do something different?

I killed the switch statement in `FadeAnimator` on my local copy.

I'll update my Pull Request, update the CHANGELOG, or possibly submit a new Pull Request for my work because it looks like there's a merge conflict.

It may be a couple more days until I can do this, I have a huge project unexpectedly taking up my time.

Finally, this project is truly amazing and it shows in the complexity of it. While design/UI is not my strong point, hopefully, I can continue to contribute by cleaning up the code where ever I can.

I think this refactor is a temporary cleanup and I think better solutions will expose themselves as I dive deeper into the project and take a look at the other `Animators`

Thanks

JakeLin commented:
@SD10 thanks, 
>in the case of .systemSuckEffect, .systemRippleEffect, .systemRotate, and .explode
>I just returned the original TransitionAnimationType
>Would you like me to do something different?

No need to have `reversed` for `.systemSuckEffect`, `.systemRippleEffect`, `.systemRotate`, and `.explode`. You are doing it right.

Can you rebase? It should not have lots of conflicts. We should merge your PR bit by bit instead of re-created bigger and bigger PR. It can help maintain conflicts.

tbaranes commented:
@SD10 Plus @JakeLin rebase suggestion, don't hesitate to commit progressively your changes since (at least when it's fixing comments), it makes review easier to do! :)

SD10 committed:
Remove Switch statement from FadeAnimator initializer

SD10 committed:
Update CHANGELOG.md for #427

SD10 committed:
Fix CHANGELOG.md

SD10 commented:
@tbaranes @JakeLin 
No problem I implemented the changes you mentioned and updated the CHANGELOG.md
As for committing progressively I can do that as well. I'm just use to working on projects where they want the smallest number of commits possible or 1 commit per pull request.
Thanks

EDIT: 
Looks like that rebase didn't go so well. It tried to do a 3 way merge and I had to resolve the .pbxproj merge conflict manually. Will have to look into this again later...

tbaranes commented:
@SD10 Thanks for doing this! For the git history, we can still squash it at the end if necessary 😬 
Can you ping us once you think your PR is ready to go or need a new review?

SD10 commented:
@tbaranes it says there are no conflicts with the base branch but the build failed
because Swift Lint can't find the `SystemAnimator` files that I deleted

I'm not sure how to fix this one

EDIT: nevermind... I know exactly what I did wrong 🤕

JakeLin commented on CHANGELOG.md:
```diff
@@ -18,6 +18,8 @@ All notable changes to this project will be documented in this file.
 [#417](https://github.com/IBAnimatable/IBAnimatable/pull/417) by [@phimage](https://github.com/phimage)
 - Add `RefreshControlDesignable` to make `UIRefreshControl` customization available in Interface Builder. Currently supported by `UITableViewController` and `UITableView`
 [#418](https://github.com/IBAnimatable/IBAnimatable/pull/418) by [@phimage](https://github.com/phimage) and [#429](https://github.com/IBAnimatable/IBAnimatable/pull/429) by [@tbaranes](https://github.com/tbaranes)
+- Replace all `SystemAnimator` classes with `SystemTransitionAnimator`
```
This is a `API breaking changes` 😁, that's the last bit we can merge the PR.

SD10 replied:
Done 👍

SD10 commented:
@tbaranes @JakeLin 
Alright, all the changes have been made and this Pull Request is complete on my end.

Sorry for all the delays. I got distracted working on this project. 
If either of you happens to use Firebase - https://github.com/SD10/Nora

JakeLin commented:
@SD10 It is in, thank you very much, well done for removing a bunch of code  🎉 🎉 🎉 🎉 🎉. Nora looks cool, we use Firebase for analytics but not for storage. Will check it out when I use it as a database.

If you would like to work on IBAnimatable more, would you like to join us as an internal contributor? We can add you to IBAnimatable GitHub organisation and Slack channel. Please let me know if you are interested. Thanks again, you make IBAniamtable better and better 💪

SD10 commented:
@JakeLin I would be honored to join your organization and I plan on contributing to IBAnimatable more. I can't say that my skills revolve around UI. I love refactoring code, optimizing performance, and creating tools for other developers. I'll contribute any way that I can.

JakeLin commented:
@SD10 Welcome 🎉🎉🎉🎉🎉 I have sent the invite to you. Can you please send me your Slack email? You can find my email on my GitHub profile.

To be honestly, I am not good at anything, I learn every day from you guys. I like to contribute whatever I can and learn from the others to improve my skills. The rule for us is simple: We don't have to work on the project, we contribute when we are insterested and have time to. We respect to each other and open to any opionions. We use feature branches like `feature/***` or `bugfix/***` and merge to `master` after one or two 👍 . We have fun together 👯

SD10 commented:
@JakeLin You're too humble. This is an awesome project!

Thanks for the invite! You can use the same email for the Slack
stevensdeutsch@yahoo.com

Thanks again

JakeLin commented:
@SD10 thanks, invite sent, let's have some fun together 👯
